drivers: display: convert st7735r to MIPI DBI API

danieldegrasse opened the pull request:
This PR converts the ST7735R to the MIPI DBI API. Boards using this display have also been updated.

Additionally, this PR adds support for a new API in the MIPI DBI driver, `mipi_dbi_release`. This API mirrors the functionality of `spi_release`, and is added because some displays depend on the `SPI_HOLD_ON_CS` functionality in order to work correctly- therefore, the MIPI DBI API needs to offer a similar ability to hold the CS signal low.


~In draft until https://github.com/zephyrproject-rtos/zephyr/pull/71454 merges~

danieldegrasse committed:
drivers: display: st7735r: convert to MIPI DBI API

Convert the ST7735R display to use the MIPI DBI API. Boards and overlays
using this display are also updated.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
tests: drivers: build_all: add st7735r to test

Add entry for st7735r to display build all test

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
docs: releases: migration-guide-3.7: add note about ST7735R migration

Add note about migration of ST7735R displays to MIPI DBI API.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@jfischer-no I know you mentioned you have most in tree SPI displays- do you have this one, and if so would you be willing to test this change? I don't have hardware to verify this with

mmahadevan108 commented on drivers/display/display_st7735r.c:
```diff
@@ -23,15 +24,14 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(display_st7735r, CONFIG_DISPLAY_LOG_LEVEL);
 
-#define ST7735R_RESET_TIME              K_MSEC(1)
```
Do we need to change this?

danieldegrasse replied:
Yes, we do- the MIPI DBI reset API takes a delay time in milliseconds, not a k_timeout_t structure. Therefore we need to pass an integer reset delay here, not `K_MSEC(1)`

dleach02 commented:
@danieldegrasse, please address the merge conflict

mmahadevan108 commented:
@danieldegrasse, can we move the first 3 commits to a separate PR as they are related to updating the MIPI-DBI API and can be looked at separately.

danieldegrasse commented:
> @danieldegrasse, please address the merge conflict

Done, thanks for the reminder

> @danieldegrasse, can we move the first 3 commits to a separate PR as they are related to updating the MIPI-DBI API and can be looked at separately.

Sure- I've created a PR here: https://github.com/zephyrproject-rtos/zephyr/pull/71454. Marking this PR as draft, since it now depends on that one

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -123,6 +123,49 @@ Controller Area Network (CAN)
 
+* ST7735R based displays now use the MIPI DBI driver class. These displays
+  must now be declared within a MIPI DBI driver wrapper device, which will
+  manage interfacing with the display. Note that the `cmd-data-gpios` pin has
```
```suggestion
  manage interfacing with the display. Note that the ``cmd-data-gpios`` pin has
```

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -123,6 +123,49 @@ Controller Area Network (CAN)
+  must now be declared within a MIPI DBI driver wrapper device, which will
+  manage interfacing with the display. Note that the `cmd-data-gpios` pin has
+  changed polarity with this update, to align better with the new
+  `dc-gpios` name. For an example, see below:
```
```suggestion
  ``dc-gpios`` name. For an example, see below:
```

kartben commented on dts/bindings/mipi-dbi/mipi-dbi-device.yaml:
```diff
@@ -11,3 +11,14 @@ properties:
+  mipi-mode:
+    type: int
+    description: |
+      MIPI DBI mode in use. Use the macros not the actual enum value, here is
+      the concordance list (see dt-bindings/mipi_dbi/mipi_dbi.h)
```
```suggestion
      MIPI DBI mode in use. Use the macros, not the actual enum value. Here is
      the concordance list (see dt-bindings/mipi_dbi/mipi_dbi.h)
```

kartben commented on include/zephyr/drivers/mipi_dbi.h:
```diff
@@ -256,6 +264,31 @@ static inline int mipi_dbi_reset(const struct device *dev, uint32_t delay)
+ * needs to keep CS asserted for multiple transactions, or the MIPI DBI device
+ * locked.
+ * @param dev mipi dbi controller
+ * @param config MIPI DBI configuration
```
missing doc for return value?

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -123,6 +123,49 @@ Controller Area Network (CAN)
 Display
 =======
 
+* ST7735R based displays now use the MIPI DBI driver class. These displays
```
you may also add a release not entry along the lines of "Updated ST7735R based displays to use the MIPI DBI driver class", and of course keep this (great) entry for migration instructions.

danieldegrasse commented:
@kartben Requested fixes for migration guide should be applied to this PR. The other documentation changes you mentioned were part of #71454 (which is merged), so I opened #71715 to apply them

kartben reviewed (dismissed):
LGTM - thanks

danieldegrasse commented:
@ozersa would you be able to look over (and potentially test) this change?

ozersa commented:
> @ozersa would you be able to look over (and potentially test) this change?

Unfortunately I do not have change to test it, As per of my review

For 3wire mode, 9bit SPI data sends, that requires below line being 2 instead of 1.

https://github.com/zephyrproject-rtos/zephyr/blob/a1cdf276101656ebfb0f3274b5771005f7ca5c78/drivers/mipi_dbi/mipi_dbi_spi.c#L78

ozersa commented on drivers/display/display_st7735r.c:
```diff
@@ -530,11 +491,10 @@ static const struct display_driver_api st7735r_api = {
-		.reset = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, {}),	\
+		.mipi_dev = DEVICE_DT_GET(DT_INST_PARENT(inst)),		\
+		.dbi_config = MIPI_DBI_CONFIG_DT_INST(inst,			\
+				SPI_OP_MODE_MASTER | SPI_WORD_SET(8) |		\
```
SPI_WORD_SET(8)
It will be better if you set correct SPI data size depend on the selected mode (8 for 4wire, 9 for 3wire).

danieldegrasse replied:
I think this makes sense. Doing so will also allow dropping the `tmp_cfg` within the MIPI DBI driver.

danieldegrasse commented:
Marking [DNM] until I address @ozersa's review

danieldegrasse committed:
drivers: mipi_dbi: mipi_dbi_spi: require 3 wire mode use 9 bit SPI

Require that SPI config within the MIPI DBI API use 9 bit SPI mode, as 3
wire spi requires the command/data bit be packed into the start of the
SPI packet.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

ozersa reviewed (approved):
Looks good.

henrikbrixandersen commented:
@danieldegrasse Please rebase on `main`.

danieldegrasse commented:
> @danieldegrasse Please rebase on main.

Done, thanks
